### ISSUE TITLE
reassembly stream not closing connections properly?

### ISSUE THREAD

**Issue by r0t3n** [google/gopacket#330]
I have made a module utilising reassembly streams to easily catch the opening, establishment and closing of connections for a statistics counter.

I am using live pcap with a filter which only catches packets with syn, fin or rst. I have run tcpdump alongside my tests and can see the FIN-ACK packets from both ends of the connection. The connections do not show up as closed (no call to ReassemblyComplete()) unless I forcefully close them via flushing. 

Is this the intended behaviour? If so, as I'm not capturing data packets as I only care about the connection state, is there a way for it to call ReassemblyComplete() in realtime without flushing other long lived connections out of the pool that I need to track.

**Comment by gconnell** [comment]
Hey, r0t3n,

My guess, without delving too deeply, is that your filter is actually what's causing you issues.  If you're using any of the reassembly code in GoPacket, that code attempts to reassemble an entire stream.  So, if it sees Packet1, <missing packets>, Packet10, it'll assume it's getting an out-of-order packet and wait around for packets 2-9.  Since you're dropping ALL packets between the SYN and FIN/RST, none of your streams are actually closing... they're all buffering the FIN/RST and waiting around for the out-of-order intermediates.

In short, ReassemblyComplete is only called if all data in the stream has been successfully reassembled (IE, all packet data has been seen), or on an explicit Flush call.

If you're really just looking at SYN and FIN/RST, you probably don't need the reassembly code at all... a simple map[] with flow endpoints will probably do all you need, since you really want just a single bit ("is connection open") for each connection.

  --Graeme
